### DISCUSSION TITLE
Species/column-species intercepts

### DISCUSSION THREAD

**Discussion by hrlai** [JenniNiku/gllvm#154] [Q&A] [answered]
Hi there, I tried to find answers from the helpfile without avail: are column-specific intercepts fitted as fixed or random terms? I am guess that they are fixed when `beta0com = FALSE` and random otherwise? I'm also guessing that they are fixed by default because a model with `beta0com = FALSE` doesn't seem to return any variance / sigma parameter for the column intercepts.

**Comment by BertvanderVeen** [comment, marked as answer]
Hello! I have converted your issue to a discussion, since you are not reporting a bug. In the future, please post questions in the "discussions" tab on github, and potential bug reports under "issues".

Column specific intercepts are treated as fixed effects at all times. `beta0com` is an option that only applies (at present) to the fourth-corner model, and collects all column intercepts to a common value, but still treats that as a fixed effect. It is possible to additionally include random intercepts in order to estimate their variance, with the development version on the package that is available on github.

It is not currently possible to omit the fixed effects intercepts entirely from the model.

**Reply by hrlai** [reply]
Thank you @BertvanderVeen . I kept forgetting that there is a Discussion section.

Alright then I'll keep this in mind. We are trying to predict the distribution of a hypothetical species with counterfactual traits, and ran into the problem of what intercept to use. If species intercepts are random with a population mean, then this is easy; but less clear when they are fixed intercepts...

**Reply by BertvanderVeen** [reply]
OK! Same principle holds for the fourth corner model: add a column of intercepts and specify in the "randomX" argument in combination with beta0comm = TRUE, that you get you what you are after.

**Reply by hrlai** [reply]
Great thanks. I tried it out with the example data and got some follow-up questions:
```
library(gllvm)
library(mvabund)
data(antTraits)
y <- as.matrix(antTraits$abund)
X <- as.matrix(antTraits$env)
TR <- antTraits$traits[, c(1, 5)]
fitF3 <- gllvm(y = y, X = scale(X), TR = scale(TR),
               formula = ~ Bare.ground + Canopy.cover * (Femur.length + Webers.length),
               family = "negative.binomial", 
               randomX = ~ Bare.ground + Canopy.cover, 
               beta0com = TRUE,
               control.start = list(n.init = 3))
```

1. `summary(fitF3)` gave me an error: 

``` 
Error in dimnames(coef.table) <- list(newnam, c("Estimate", "Std. Error",  : 
  length of 'dimnames' [1] not equal to array extent
In addition: Warning messages:
1: In pars/se :
  longer object length is not a multiple of shorter object length
2: In cbind(pars, se, zval, pvalue) :
  number of rows of result is not a multiple of vector length (arg 1) 
```

2. `fitF3$params` now gives `beta0` the common intercept as a single value, which is nice. But there is no species-specific intercepts, do they need to be calculated as something like `beta0 + B * traits` manually?
3. In `sigmaB` I can see the variance-covariance (or are they SDs?) of `B`, but there is no variance for the common intercept...? Is this related to what you mentioned at first "...still treats that as a fixed effect" ?

Thanks!

**Reply by BertvanderVeen** [reply]
Thanks. I made a few adjustments in the development branch for this, also in the summary method, which should fix your issue (I tried, it worked fine).

What I meant before, is that you need to explicitly include an intercept column in the model, and pass it to the "randomX" argument:

```
library(gllvm)
library(mvabund)
data(antTraits)
y <- as.matrix(antTraits$abund)
X <- as.matrix(antTraits$env)
X <- cbind(int = 1, X)
TR <- antTraits$traits[, c(1, 5)]
fitF3 <- gllvm(y = y, X = X, TR = scale(TR),
               formula = ~ Bare.ground + Canopy.cover * (Femur.length + Webers.length),
               family = "negative.binomial", 
               randomX = ~ int + Bare.ground + Canopy.cover, 
               beta0com = TRUE,
               control.start = list(n.init = 3))
```

unfortunately, there seems to be a bug in the package that does not at present allow you to do this in the fourth corner model. I have pushed a commit (https://github.com/JenniNiku/gllvm/commit/009484a6b7a201f862a6439af138e4f6fe8d0304), which should address this, albeit it is probably not the most pretty fix. @JenniNiku tagging you because you might want to review the change as well.

Please reinstall the package directly from the repository and try again.

**Reply by hrlai** [reply]
Ah, thanks for this! I can also confirm that `summary` not works perfectly. 

For now nothings seems to break with your latest commit.

**Reply by hrlai** [reply]
Oh actually one minor thing that isn't urgent: with your latest example `fitF3$params$beta0` returns species-specific values that are identical --- I presume that they are just the single community intercept.

**Reply by BertvanderVeen** [reply]
Yes, that is correct. I quickly re-coded how these are handled, so `model$params$beta0` is always a species-length vector for convenience.
